### PR TITLE
couch to sql migration / prefactor

### PR DESCRIPTION
**Pull request by czue**
@dimagi/scale-team this isn't done yet but wanted to get some high-level feedback on the approach and also run tests. 

I think this is going to be quite doable though the current approach I'm taking involves forking a decent amount of the form-processing code which I can see being a pain to maintain over time. The other option would be to add parameters to those methods and keep the code in one place which adds a different kind of complexity.

I jotted this down, as well as a number of other thoughts/questions in [the google doc](https://docs.google.com/document/d/1eIU6pJ6xBQokCa7AHUxOWDUFX2kLfwl6kIoNuRRpcts/edit#heading=h.a4ng5uahq1ik).

Probably useful to read :fish: though it may make sense to review `migrate_domain_from_couch_to_sql.py` all at once

### CODE REVIEW

File: corehq/form_processor/management/commands/migrate_domain_from_couch_to_sql.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
+from datetime import datetime
+from django.core.management.base import LabelCommand, CommandError
+from casexml.apps.case.xform import get_all_extensions_to_close, CaseProcessingResult
+from corehq.form_processor.interfaces.processor import FormProcessorInterface
+from corehq.form_processor.models import XFormInstanceSQL
+from corehq.form_processor.submission_post import CaseStockProcessingResult
+from corehq.form_processor.utils import should_use_sql_backend, adjust_datetimes
```
Comment (dimagimon) [inline corehq/form_processor/management/commands/migrate_domain_from_couch_to_sql.py]: F401 'adjust_datetimes' imported but unused

File: corehq/form_processor/utils/general.py [hunk 2]
```diff
@@ -19,6 +36,11 @@ def should_use_sql_backend(domain_object_or_name):
 
     if domain_object is None:
         return False
+
+    local_override = get_local_domain_sql_backend_override(domain_name)
+    if local_override is not None:
+        return local_override
```
Comment (snopoke) [inline corehq/form_processor/utils/general.py]: nice

File: corehq/form_processor/utils/general.py [hunk 3]
```diff
@@ -29,6 +29,8 @@ def should_use_sql_backend(domain_object_or_name):
 
 
 def should_use_sql_backend_in_prod(domain_object_or_name):
+    # the only reason this method is "public" is so it can be called directly by a handful of tests.
```
Comment (benrudolph) [inline corehq/form_processor/utils/general.py]: in the past i've still kept functions prefixed with `_` even if they were just used in tests. i'm not sure if that's the correct thing to do and might say something about the level of testing

Comment (czue) [inline corehq/form_processor/utils/general.py]: yeah i debated that for a while. i think i like it with the `_`, will change

**Comment by czue** [comment]
Note that I think this is fine to merge whenever if we think the case processing refactors are net positive and trust the tests. I'll keep working on it in my spare time but not sure how much of that I'll have for the rest of the week. 

(also fine to wait as this isn't pressing at all)

**Comment by benrudolph** [comment]
i think nice to get it in and not have it bit rot

**Pull request merged**